src/tensor_ops/conv2d/conv2d.cu

```diff
@@ -78,43 +79,50 @@ __device__ void unfold_output_into_patches(
     idx /= op.w_in;
     const size_t y = idx % op.h_in;
     idx /= op.h_in;
-    const size_t k2 = idx % op.kernel;
-    idx /= op.kernel;
-    const size_t k1 = idx % op.kernel;
-    idx /= op.kernel;
     const size_t o = idx % op.chan_out;
     idx /= op.chan_out;
     const size_t b = idx % op.batch;
     idx /= op.batch;
 
-    size_t oh = y + op.padding;
-    if (oh < k1) {
-        return;
-    }
-    oh -= k1;
-    if (oh % op.stride != 0) {
-        return;
+    T img = image_out[i];
+
+    for (int k1=0; k1<op.kernel; k1++) {
+        for (int k2=0; k2<op.kernel; k2++) {
+            size_t oh = y + op.padding;
+            if (oh < k1) {
+                return;
+            }
+            oh -= k1;
+            if (oh % op.stride != 0) {
+                return;
+            }
+            oh /= op.stride;
+            if (oh >= op.h_out) {
+                return;
+            }
+            
+            size_t ow = x + op.padding;
+            if (ow < k2) {
+                return;
+            }
+            ow -= k2;
+            if (ow % op.stride != 0) {
+                return;
+            }
+            ow /= op.stride;
+            if (ow >= op.w_out) {
+                return;
+            }
+
+            size_t patch_i = b * (op.chan_out * op.kernel * op.kernel * op.h_out * op.w_out) + 
```

Comment:
Some of this calculation could be moved outside the loop, but if its optimized right I don't think it would make a difference.